golos.ctrl/src/golos.ctrl.cpp

```diff
@@ -177,25 +177,35 @@ void control::unregwitness(name witness) {
     assert_started();
     require_auth(witness);
 
-    // TODO: simplify upsert to allow passing just inner lambda
-    bool exists = upsert_tbl<witness_tbl>(witness, [&](bool) {
-        return [&](witness_info& w) {
-            eosio_assert(w.active, "witness already unregistered");
-            w.active = false;
-        };
-    }, false);
-    eosio_assert(exists, "witness not found");
+    witness_tbl witness_table(_self, _self.value);
+    auto it = witness_table.find(witness.value);
+    witness_table.erase(*it);
 
+    //TODO remove votes for witness
     update_auths();
 }
 
+void control::stopwitness(name witness) {
+    assert_started();
+    require_auth(witness);
+    active_witness(witness, false);
+}
+
+void control::startwitness(name witness) {
+    assert_started();
+    require_auth(witness);
+    active_witness(witness, true);
```

Comment:
It will be useful to send notification about witness state when witness activates/deactivates. You can add witness active flag in exist notification.